### ISSUE TITLE
OPTIONAL UNWIND

### ISSUE THREAD

**Issue by jmarton** [opencypher/openCypher#234]
# CIR-2017-234

Allow for the `OPTIONAL` modifier in `UNWIND` to retain those patterns with an empty list. The idea is similar to the `OPTIONAL MATCH` construct in openCypher, or the *left outer join* in SQL.

# Overview

`UNWIND` allows multiplying a pattern match by substiuting list elements in place of the list itself.

I.e. given 4 person nodes along with the languages they command:

```
create (:Demo {name: 'Alice', languages: ['en', 'de', 'gr']})
     , (:Demo {name: 'Bob', languages: ['en', 'de']})
     , (:Demo {name: 'Cecil', languages: []})
     , (:Demo {name: 'Dennis'})
```

`UNWIND`ing languages results in 3 matches for Alice and 2 for Bob. However, Cecil is hidden from the output because the associated languages list is the empty list. Dennis is also hidden as he has no languages property at all:

```
MATCH (n:Demo)
UNWIND n.languages AS lang
RETURN n.name, lang
```

yields:

```
╒════════╤══════╕
│"n.name"│"lang"│
╞════════╪══════╡
│"Alice" │"en"  │
├────────┼──────┤
│"Alice" │"de"  │
├────────┼──────┤
│"Alice" │"gr"  │
├────────┼──────┤
│"Bob"   │"en"  │
├────────┼──────┤
│"Bob"   │"de"  │
└────────┴──────┘
```

## The problem

Note that in the example above, Cecil was hidden from the final resultset as he commands no languages.

This is sometimes undesirable, and it might be beneficial to retain a record in the resultset for him, giving NULL as the lang.

# Request

Allow the `OPTIONAL` modifier for the UNWIND clause which would retain those matches with an empty list, but suppress those that has no such attribute defined.

## Syntax example

```
MATCH (n:Demo)
OPTIONAL UNWIND n.languages AS lang
RETURN n.name, lang
```

should yield the following by showing Cecil with `lang=NULL`, but still suppressing Dennis:

```
╒════════╤══════╕
│"n.name"│"lang"│
╞════════╪══════╡
│"Alice" │"en"  │
├────────┼──────┤
│"Alice" │"de"  │
├────────┼──────┤
│"Alice" │"gr"  │
├────────┼──────┤
│"Bob"   │"en"  │
├────────┼──────┤
│"Bob"   │"de"  │
├────────┼──────┤
│"Cecil" │NULL  │
└────────┴──────┘
```

# To be determined

If it makes more behaviour more consistent, `OPTIONAL UNWIND` might also retain those pattern matches that has no such property defined, i.e. for the example given, 

```
╒════════╤══════╕
│"n.name"│"lang"│
╞════════╪══════╡
│"Alice" │"en"  │
├────────┼──────┤
│"Alice" │"de"  │
├────────┼──────┤
│"Alice" │"gr"  │
├────────┼──────┤
│"Bob"   │"en"  │
├────────┼──────┤
│"Bob"   │"de"  │
├────────┼──────┤
│"Cecil" │NULL  │
├────────┼──────┤
│"Dennis"│NULL  │
└────────┴──────┘
```

**Comment by thobe** [comment]
I really like this idea.

**Comment by Mats-SX** [comment]
This seems useful. We should try to align the proposed semantics with the [UNWIND CIP](https://github.com/opencypher/openCypher/pull/230), which suggests that `null` is just another value for `UNWIND`, and is treated like a single-element list, thus modifying the first result table in this CIR's description.

Given that, the only difference between `UNWIND` and `OPTIONAL UNWIND` would be the treatment of empty lists. Unfortunately I do not see a way in which `OPTIONAL UNWIND` would be able to treat empty lists and `null`s different in this context, but perhaps that can be considered outside the scope of this CIR.

**Comment by thobe** [comment]
I actually think it is at least worth considering whether `UNWIND` and `OPTIONAL UNWIND` should both treat `null` and `[]` the same. I.e. what has been proposed in the [UNWIND CIP](https://github.com/opencypher/openCypher/pull/230) should be possible to change in the light of `OPTIONAL UNWIND`.

**Comment by Mats-SX** [comment]
That idea is actually more in alignment (as in: less breakage) with the current Neo4j implementation, where `null` _is_ treated the same as the empty list. It is less in alignment with the somewhat prevalent 'null-in-null-out' style of many other Cypher constructs.

I think both approaches have merit; it comes down to considering what `null` sensibly means in this context: is it more like a single missing/unknown value (--> singleton list interpretation), or more like a list with missing/unknown values (--> empty list interpretation)?

**Comment by jmarton** [comment]
I agree to treat `[]` and `null` the same in either of `UNWIND` and `OPTIONAL UNWIND`.

Can we say that there is a consensus on that `[]` as well as `null`:

 - should zero out cardinality for `UNWIND`
 - should retain cardinality of one for `OPTIONAL UNWIND`

However, I would not suggest that for an attribute of list semantics, the `null` value and the `[] ` empty list mean the same.

**Comment by Mats-SX** [comment]
@jmarton What do you mean with the last paragraph?

**Comment by jmarton** [comment]
@Mats-SX you wrote:

> what `null` sensibly means in this context: *[...]* or more like a list with missing/unknown values (--> empty list interpretation)?

I just meant that if we say `null` in this context is a list with missing/unknown values, it is not the *empty list*. I'd rather say it's an *arbitrary list*.

**Comment by Mats-SX** [comment]
Is there a particular point in making that distinction that I'm missing? The proposed semantics of `[OPTIONAL] UNWIND` would treat `null` identical to `[]`, but different from any other list. Isn't that the same as making `[OPTIONAL] UNWIND` assume `null` is empty?

**Comment by jmarton** [comment]
For the `[OPTIONAL] UNWIND` operations, `null` and `[]` are the same, i.e. they treat them the same way.

It's just about wording. I just wanted to point out that it is unfortunate wording to say *a list with missing/unknown values (--> empty list interpretation)*, and it seems more clean for me to say *a null in list position means an arbitrary/unknown list*.

**Comment by quiin** [comment]
Are there any updates on this matter?
I think `OPTIONAL UNWIND` would be useful to have considering this example:
```
MERGE(main_node: Case {uid: 1})
UNWIND $relations_out as relation_to
CALL apoc.merge.node([relation_to.node], relation_to.identifiers, null) YIELD node as to_node
CALL apoc.merge.relationship(main_node, relation_to.label, null, null, to_node) YIELD rel as rel_to
UNWIND $relations_in as relation_from
CALL apoc.merge.node([relation_from.node], relation_from.identifiers, null) YIELD node as from_node
CALL apoc.merge.relationship(from_node, relation_from.label, null, null, main_node) YIELD rel as rel_from
return main_node
```

On the example above, the second `UNWIND` won't be executed if `$relations_out` is empty

**Comment by aonamrata** [comment]
This is such an old request. Anyone has a workaround this?

**Comment by szarnyasg** [comment]
@aonamrata - it's possible to work around this using a [`CASE` statement](https://neo4j.com/docs/cypher-manual/current/syntax/expressions/#syntax-simple-case) that checks whether the given property exists and whether its an empty list. There are multiple workarounds depending on which lines you want to keep.
I'm using the example in @jmarton's original post.

```
OPTIONAL MATCH (n:Demo)
UNWIND
  CASE
    WHEN exists(n.languages) THEN n.languages
    ELSE [null]
  END AS lang
RETURN n.name, lang
```
Keeps Dennis but not Cecil.

```
OPTIONAL MATCH (n:Demo)
UNWIND
  CASE
    WHEN size(n.languages) > 0 THEN n.languages
    ELSE [null]
  END AS lang
RETURN n.name, lang
```
Keeps both Cecil and Dennis.

```
OPTIONAL MATCH (n:Demo)
UNWIND
  CASE
    WHEN size(n.languages) > 0 THEN n.languages
    WHEN size(n.languages) = 0 THEN [null]
    ELSE []
  END AS lang
RETURN n.name, lang
```
Keeps Cecil but not Dennis.

**Comment by aonamrata** [comment]
Ah in my case I cannot use `[null]` as match won't work . 
In this row is from the yield. So when I do something like :point_down: 

```
MERGE (pNew:Profile { profileId: row.id, profileType: 'InsightsProfile' }) 

UNWIND split(row.vendorIds, ',') as vendorId
    MATCH (v:Vendor { id: toInteger(vendorId) })
    MERGE (p)-[rel:HAS_ACCESS_TO]->(v)
WITH p, row

UNWIND split(row.subaccountIds, ',') as subaccountId
    MATCH (s:SubAccount { id: toInteger(subaccountId) })
    MERGE (p)-[rel:HAS_ACCESS_TO]->(s)

```
It wont generate `SubAccount` relation for users that dont have vendorIds.

**Comment by szarnyasg** [comment]
Okay, I see.

Question: are `p` and `pNew` supposed to be the same variable?

I think `+ [null]` could do the trick in your case, just make sure that the computation passes the `WITH` barrier by e.g. stating `WITH p, row, count(*)`. The count aggregation will cause the query to continue even when there are 0 matches.

Another idea: maybe as a workaround specific to this case, would it be possible to swap the `UNWIND` operations? This could work if `subaccountIds` is always non-empty.
